Mathlib/Tactic/TermCongr.lean

```diff
@@ -0,0 +1,695 @@
+  let pf ← Term.elabTerm h none
+  let pfTy ← inferType pf
+  unless ← isDefEq (← inferType pfTy) (.sort .zero) do
+    throwError "Hole has type{indentD pfTy}\nbut is expected to be a Prop"
+  if let some (_, lhs, _, rhs) ← sides? pfTy then
+    let val := if forLhs then lhs else rhs
+    if let some expectedType := expectedType? then
+      -- Propagate type hint:
+      discard <| isDefEq expectedType (← inferType val)
+    mkCHole forLhs val pf
+  else
+    -- Since `pf` doesn't yet have sides, we resort to the value and the proof being decoupled.
+    -- These will be unified during congruence generation.
+    mkCHole forLhs (← mkFreshExprMVar expectedType?) pf
+
+/-- (Internal for `congr(...)`)
+Elaborates to an expression satisfying `cHole?` that equals the LHS or RHS of `h`,
+if the LHS or RHS is available after elaborating `h`. Uses the expected type as a hint. -/
+syntax (name := cHoleExpand) "cHole% " (&"lhs" <|> &"rhs") term : term
+
+@[term_elab cHoleExpand]
+def elabCHoleExpand : Term.TermElab := fun stx expectedType? =>
+  match stx with
+  | `(cHole% lhs $h) => elabCHole h true expectedType?
+  | `(cHole% rhs $h) => elabCHole h false expectedType?
+  | _ => throwUnsupportedSyntax
+
+/-- Replace all `term` antiquotations in a term using the given `expand` function. -/
+def processAntiquot (t : Term) (expand : Term → Term.TermElabM Term) : Term.TermElabM Term := do
+  let t' ← t.raw.replaceM fun s => do
+    if s.isAntiquots then
+      let ks := s.antiquotKinds
+      unless ks.any (fun (k, _) => k == `term) do
+        throwErrorAt s "Expecting term"
+      let h : Term := ⟨s.getCanonicalAntiquot.getAntiquotTerm⟩
+      expand h
+    else
+      pure none
+  return ⟨t'⟩
+
+/-- Given the pattern `t` in `congr(t)`, elaborate it for the given side
+by relpacing antiquotations with `cHole%` terms, and ensure the elaborated term
+is of the expected type. -/
+def elaboratePattern (t : Term) (expectedType? : Option Expr) (forLhs : Bool) :
+    Term.TermElabM Expr :=
+  Term.withoutErrToSorry do
+    let t' ← processAntiquot t (fun h => if forLhs then `(cHole% lhs $h) else `(cHole% rhs $h))
+    Term.elabTermEnsuringType t' expectedType?
+
+/-! ### Congruence generation -/
+
+/-- Generates a congruence lemma for a function `f` and `numArgs` of its arguments.
+The only `Lean.Meta.CongrArgKind` kinds that appear in such a lemma
+are `.eq`, `.heq`, and `.subsingletonInst`.
+The resulting lemma proves either an `Eq` or a `HEq` depending on whether the types
+of the LHS and RHS are equal or not.
+
+This function is a wrapper around `Lean.Meta.mkHCongrWithArity`.
+It transforms the resulting congruence lemma by trying to automatically prove hypotheses
+using subsingleton lemmas, and if they are so provable they are recorded with `.subsingletonInst`.
+Note that this is slightly abusing this kind since (1) it might not be for a `Decidable` instance
+and (2) it might not even be for an instance. -/
```

Comment:
I couldn't follow this sentence.

Reply:
Added clarification. "Kind" refers to a `CongrKind`, and this function uses `subsingletonInst` for arguments that are subsingletons for reasons other than being `Decidable` instances.